activerecord/lib/active_record/insert_all.rb

```diff
@@ -116,9 +126,10 @@ def returning
         end
 
         def conflict_target
-          return unless conflict_columns
-          sql = +"(#{quote_columns(conflict_columns).join(',')})"
-          sql << " WHERE #{where}" if where
+          index = insert_all.unique_by_index
+
+          sql = +"(#{quote_columns(index&.columns || insert_all.primary_keys).join(',')})"
```

Comment:
We can also provide the name of an index as **conflict target**! so:

> Example:
> ```sql
> INSERT ... ON CONFLICT index_books_on_isbn DO UPDATE ...
> ```

If we _only_ support identifying indexes by name, I'm pretty sure this method boils down to

```ruby
def conflict_target
  insert_all.unique_by_index
end
```

Reply:
<img width="410" alt="Screenshot 2019-03-09 at 20 48 51" src="https://user-images.githubusercontent.com/350807/54076518-db25b380-42ac-11e9-8b42-a7d7d3bea75b.png">

Reply:
Hm, I cannot seem to get that to work. And I can't find documentation for either SQLIte3, Postgres or MySQL that say they support it. How does it work?

Reply:
🤔 I might've read https://www.postgresql.org/docs/9.5/sql-insert.html#SQL-ON-CONFLICT wrong. But does it work like this? (with `ON CONSTRAINT`)

```sql
INSERT ... ON CONFLICT ON CONSTRAINT index_books_on_isbn DO UPDATE ...
```

Then again, maybe that's not the best plan:

> Tip: It is often preferable to use unique index inference rather than naming a constraint directly using `ON CONFLICT ON CONSTRAINT constraint_name`. Inference will continue to work correctly when the underlying index is replaced by another more or less equivalent index in an overlapping way, for example when using `CREATE UNIQUE INDEX ... CONCURRENTLY` before dropping the index being replaced.

Reply:
I tried that but it threw an "Unknown constraint error".

Ah nice tidbit! So that's why they prefer to infer the index instead of naming it.